src/geometry/manifold/basic_smooth_bundle.lean

```diff
@@ -0,0 +1,568 @@
+this kind of bundle, that we call basic smooth bundles.
+
+A basic smooth bundle is thus a smooth vector bundle over a smooth manifold which is trivial in
+coordinate charts. It can be constructed from a basic smooth bundled core, defined below, specifying
+the changes in the fiber when one goes from one coordinate chart to another one.
+
+## Main definitions
+
+`basic_smooth_bundle_core I M F`: M is a smooth manifold over the model with corners I on (𝕜, E, H),
+       and F is a normed vector space over 𝕜. This structure registers, for each pair of charts
+       of M, a smooth change of coordinates on F. This is the core structure from which one will
+       build a smooth vector bundle with fiber F over M.
+Let `Z` be a basic smooth bundle core over M with fiber F. We define
+`Z.to_topological_fiber_bundle_core`, the (topological) fiber bundle core associated to Z. From it,
+we get a space `Z.to_topological_fiber_bundle_core.total_space` (which as a Type is just M × F),
+with the fiber bundle topology. It inherits a manifold structure (where the charts are in bijection
+with the charts of the basis). We show that this manifold is smooth.
+
+Then we use this machinery to construct the tangent bundle of a smooth manifold.
+
+* `tangent_bundle_core I M`: the basic smooth bundle core associated to a smooth manifold M over a
+                             model with corners I.
+* `tangent_bundle I M`     : the total space of `tangent_bundle_core I M`. It is itself a
+                             smooth manifold over the model with corners `I.tangent`, the product of
+                             `I` and the trivial model with corners on `E`.
+* `tangent_space I x`      : the tangent space to M at x
+* `tangent_bundle.proj I M`: the projection from the tangent bundle to the base manifold
+
+## TODO
+construct the cotangent bundle, and the bundles of differential forms. They should follow
+functorially from the description of the tangent bundle as a basic smooth bundle.
+
+## Tags
+Smooth fiber bundle, vector bundle, tangent space, tangent bundle
+-/
+
+noncomputable theory
+
+universe u
+
+open topological_space set
+
+/-- Core structure used to create a smooth bundle above M (a manifold over the model with
+corner I) with fiber the normed vector space F over 𝕜, which is trivial in the chart domains of M.
+This structure registers the changes in the fibers when one changes coordinate charts in the base. -/
+structure basic_smooth_bundle_core {𝕜 : Type*} [nondiscrete_normed_field 𝕜]
+{E : Type u} [normed_group E] [normed_space 𝕜 E]
+{H : Type*} [topological_space H] (I : model_with_corners 𝕜 E H)
+(M : Type*) [topological_space M] [manifold H M] [smooth_manifold_with_corners I M]
+(F : Type u) [normed_group F] [normed_space 𝕜 F] :=
+(coord_change      : atlas H M → atlas H M → H → F → F)
+(coord_change_self :
+  ∀i : atlas H M, ∀ x ∈ i.1.target, ∀v, coord_change i i x v = v)
```

Comment:
I prefer leaving a space after `∀` but I guess mathlib is inconsistent here. Those `i.1.target` are really hard to read. What `.1` means here? Is it the `subtype.val` coming from the fact you see `atlas H M` as a type when defining the `coord_change` field? I guess so. Isn't there a trick we could use to get rid of all those `.1`?

Reply:
You can see on this line that I am not myself coherent with the spacing. I have added a space after all the foralls in this file. As for the `.1`, yes, it is the `subtype.val`. Maybe Lean 4 will improve things in this respect, but for now I don't know how to eliminate them.